// @szhsin/react-menu/dist/index.modern.js
import React, { useCallback, useState, useMemo, useRef, useContext, useEffect, useReducer, useLayoutEffect } from 'react';
import ReactDOM, { unstable_batchedUpdates } from 'react-dom';
import PropTypes from 'prop-types';

const menuContainerClass = 'rc-menu-container';
const menuClass = 'rc-menu';
const menuButtonClass = 'rc-menu-button';
const menuArrowClass = 'arrow';
const menuItemClass = 'item';
const menuDividerClass = 'divider';
const menuHeaderClass = 'header';
const menuGroupClass = 'group';
const subMenuClass = 'submenu';
const radioGroupClass = 'radio-group';
const initialHoverIndex = -1;
const MenuListItemContext = /*#__PURE__*/React.createContext({});
const MenuListContext = /*#__PURE__*/React.createContext({});
const EventHandlersContext = /*#__PURE__*/React.createContext({});
const RadioGroupContext = /*#__PURE__*/React.createContext({});
const SettingsContext = /*#__PURE__*/React.createContext({});
const ItemSettingsContext = /*#__PURE__*/React.createContext({});
const Keys = Object.freeze({
  'ENTER': 'Enter',
  'ESC': 'Escape',
  'SPACE': ' ',
  'HOME': 'Home',
  'END': 'End',
  'LEFT': 'ArrowLeft',
  'RIGHT': 'ArrowRight',
  'UP': 'ArrowUp',
  'DOWN': 'ArrowDown'
});
const HoverIndexActionTypes = Object.freeze({
  'RESET': 'HOVER_INDEX_RESET',
  'SET': 'HOVER_INDEX_SET',
  'UNSET': 'HOVER_INDEX_UNSET',
  'INCREASE': 'HOVER_INDEX_INCREASE',
  'DECREASE': 'HOVER_INDEX_DECREASE',
  'FIRST': 'HOVER_INDEX_FIRST',
  'LAST': 'HOVER_INDEX_LAST'
});
const SubmenuActionTypes = Object.freeze({
  'OPEN': 'SUBMENU_OPEN',
  'CLOSE': 'SUBMENU_CLOSE'
});
const CloseReason = Object.freeze({
  'CLICK': 'click',
  'CANCEL': 'cancel',
  'BLUR': 'blur',
  'SCROLL': 'scroll'
});

const batchedUpdates = unstable_batchedUpdates || (callback => callback());
const defineName = (component, name) => name ? Object.defineProperty(component, '_reactMenu', {
  value: name,
  writable: false
}) : component;
const getName = component => component && component['_reactMenu'];
const applyHOC = HOC => (...args) => defineName(HOC(...args), getName(args[0]));
const applyStatics = sourceComponent => wrappedComponent => defineName(wrappedComponent, getName(sourceComponent));
const safeCall = (fn, ...args) => typeof fn === 'function' ? fn(...args) : fn;
const attachHandlerProps = (handlers, props) => {
  const result = {};

  for (const handlerName of Object.keys(handlers)) {
    const handler = handlers[handlerName];
    const propHandler = props[handlerName];
    let attachedHandler;

    if (typeof propHandler === 'function') {
      attachedHandler = e => {
        propHandler(e);
        handler(e);
      };
    } else {
      attachedHandler = handler;
    }

    result[handlerName] = attachedHandler;
  }

  return result;
};
const parsePadding = paddingStr => {
  if (typeof paddingStr !== 'string') return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  };
  const padding = paddingStr.trim().split(/\s+/, 4).map(parseFloat);
  const top = !isNaN(padding[0]) ? padding[0] : 0;
  const right = !isNaN(padding[1]) ? padding[1] : top;
  return {
    top,
    right,
    bottom: !isNaN(padding[2]) ? padding[2] : top,
    left: !isNaN(padding[3]) ? padding[3] : right
  };
};
const bem = (block, element, modifiers = {}) => (userClassName, userModifiers) => {
  let blockElement = element ? `${block}__${element}` : block;
  let className = blockElement;

  for (const name of Object.keys(modifiers)) {
    const value = modifiers[name];

    if (value) {
      className += ` ${blockElement}--`;
      className += value === true ? name : `${name}-${value}`;
    }
  }

  if (typeof userClassName === 'function') {
    userClassName = userClassName(userModifiers || modifiers);
  }

  if (typeof userClassName === 'string') {
    userClassName = userClassName.trim();
    if (userClassName) className += ` ${userClassName}`;
  }

  return className;
};

const isObject = obj => obj && typeof obj === 'object';

const sanitiseKey = key => key.charAt(0) === '$' ? key.slice(1) : key;

const flatStyles = (styles, modifiers) => {
  if (typeof styles === 'function') return styles(modifiers);
  if (!isObject(styles)) return undefined;
  if (!modifiers) return styles;
  const style = {};

  for (const prop of Object.keys(styles)) {
    const value = styles[prop];

    if (isObject(value)) {
      const modifierValue = modifiers[sanitiseKey(prop)];

      if (typeof modifierValue === 'string') {
        for (const nestedProp of Object.keys(value)) {
          const nestedValue = value[nestedProp];

          if (isObject(nestedValue)) {
            if (sanitiseKey(nestedProp) === modifierValue) {
              Object.assign(style, nestedValue);
            }
          } else {
            style[nestedProp] = nestedValue;
          }
        }
      } else if (modifierValue) {
        Object.assign(style, value);
      }
    } else {
      style[prop] = value;
    }
  }

  return style;
};
const getScrollAncestor = node => {
  while (node && node !== document.body) {
    const {
      overflow,
      overflowX,
      overflowY
    } = getComputedStyle(node);
    if (/auto|scroll|overlay|hidden/.test(overflow + overflowY + overflowX)) return node;
    node = node.parentNode;
  }

  return window;
};
const floatEqual = (a, b, diff = 0.0001) => Math.abs(a - b) < diff;
const isProd = process.env.NODE_ENV === 'production';

const validateChildren = (parent, child, permitted) => {
  if (!child) return false;

  if (!permitted.includes(getName(child.type))) {
    !isProd && console.warn(`[react-menu] ${child.type || child} is ignored.\n`, `The permitted children inside a ${parent} are ${permitted.join(', ')}.`, 'If you create HOC of these components, you can use the applyHOC or applyStatics helper, see more at: https://szhsin.github.io/react-menu/docs#utils-apply-hoc');
    return false;
  }

  return true;
};

const cloneChildren = (children, startIndex = 0) => {
  let index = startIndex;
  let descendOverflow = false;
  const permittedChildren = ['MenuDivider', 'MenuGroup', 'MenuHeader', 'MenuItem', 'FocusableItem', 'MenuRadioGroup', 'SubMenu'];
  const items = React.Children.map(children, child => {
    if (!validateChildren('Menu or SubMenu', child, permittedChildren)) return null;

    switch (getName(child.type)) {
      case 'MenuDivider':
      case 'MenuHeader':
        return child;

      case 'MenuRadioGroup':
        {
          const permittedChildren = ['MenuItem'];
          const props = {
            type: 'radio'
          };
          const radioItems = React.Children.map(child.props.children, radioChild => {
            if (!validateChildren('MenuRadioGroup', radioChild, permittedChildren)) return null;
            return radioChild.props.disabled ? /*#__PURE__*/React.cloneElement(radioChild, props) : /*#__PURE__*/React.cloneElement(radioChild, { ...props,
              index: index++
            });
          });
          return /*#__PURE__*/React.cloneElement(child, {
            children: radioItems
          });
        }

      case 'MenuGroup':
        {
          const {
            items,
            endIndex,
            descendOverflow: descOverflow
          } = cloneChildren(child.props.children, index);
          index = endIndex;
          const takeOverflow = Boolean(child.props.takeOverflow);
          if (!isProd && (descendOverflow === descOverflow ? descOverflow : takeOverflow)) throw new Error('[react-menu] Only one MenuGroup in a menu is allowed to have takeOverflow prop.');
          descendOverflow = descendOverflow || descOverflow || takeOverflow;
          return /*#__PURE__*/React.cloneElement(child, {
            children: items
          });
        }

      default:
        return child.props.disabled ? child : /*#__PURE__*/React.cloneElement(child, {
          index: index++
        });
    }
  });
  return {
    items,
    endIndex: index,
    descendOverflow
  };
};

const stylePropTypes = name => ({
  [name ? `${name}ClassName` : 'className']: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  [name ? `${name}Styles` : 'styles']: PropTypes.oneOfType([PropTypes.object, PropTypes.func])
});
const sharedMenuPropTypes = { ...stylePropTypes(),
  ...stylePropTypes('arrow'),
  'aria-label': PropTypes.string,
  arrow: PropTypes.bool,
  offsetX: PropTypes.number,
  offsetY: PropTypes.number,
  align: PropTypes.oneOf(['start', 'center', 'end']),
  direction: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
  position: PropTypes.oneOf(['auto', 'anchor', 'initial']),
  overflow: PropTypes.oneOf(['auto', 'visible', 'hidden']),
  children: PropTypes.node.isRequired
};
const menuPropTypesBase = { ...sharedMenuPropTypes,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  animation: PropTypes.bool,
  debugging: PropTypes.bool,
  boundingBoxRef: PropTypes.object,
  boundingBoxPadding: PropTypes.string,
  reposition: PropTypes.oneOf(['auto', 'initial']),
  repositionFlag: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  viewScroll: PropTypes.oneOf(['auto', 'close', 'initial']),
  submenuOpenDelay: PropTypes.number,
  submenuCloseDelay: PropTypes.number,
  portal: PropTypes.bool,
  theming: PropTypes.string,
  onClick: PropTypes.func
};
const sharedMenuDefaultProp = {
  offsetX: 0,
  offsetY: 0,
  align: 'start',
  direction: 'bottom',
  position: 'auto',
  overflow: 'visible'
};
const menuDefaultPropsBase = { ...sharedMenuDefaultProp,
  animation: true,
  reposition: 'auto',
  viewScroll: 'initial',
  submenuOpenDelay: 300,
  submenuCloseDelay: 150
};

const useActiveState = (isHovering, isDisabled, customKey) => {
  const [active, setActive] = useState(false);
  const activeKeys = useMemo(() => [Keys.SPACE, Keys.ENTER, ...(customKey ? [customKey] : [])], [customKey]);
  const cancelActive = useCallback(() => setActive(false), []);
  return {
    isActive: active,
    onPointerDown: useCallback(() => {
      if (!isDisabled) setActive(true);
    }, [isDisabled]),
    onPointerUp: cancelActive,
    onPointerLeave: cancelActive,
    onKeyDown: useCallback(e => {
      if (isHovering && !isDisabled && activeKeys.includes(e.key)) {
        setActive(true);
      }
    }, [isHovering, isDisabled, activeKeys]),
    onKeyUp: useCallback(e => {
      if (activeKeys.includes(e.key)) {
        setActive(false);
      }
    }, [activeKeys]),
    onBlur: useCallback(e => {
      if (!e.currentTarget.contains(e.relatedTarget)) {
        setActive(false);
      }
    }, [])
  };
};

const useItemState = (isDisabled, index) => {
  const ref = useRef(null);
  const {
    submenuCloseDelay
  } = useContext(ItemSettingsContext);
  const {
    isParentOpen,
    hoverIndex,
    isSubmenuOpen,
    dispatch
  } = useContext(MenuListItemContext);
  const isHovering = hoverIndex === index;
  const timeoutId = useRef();
  const setHover = useCallback(() => {
    if (!isDisabled) dispatch({
      type: HoverIndexActionTypes.SET,
      index
    });
  }, [isDisabled, dispatch, index]);
  const onBlur = useCallback(e => {
    if (!e.currentTarget.contains(e.relatedTarget)) {
      dispatch({
        type: HoverIndexActionTypes.UNSET,
        index
      });
    }
  }, [dispatch, index]);
  const onMouseEnter = useCallback(() => {
    if (isSubmenuOpen) {
      timeoutId.current = setTimeout(setHover, submenuCloseDelay);
    } else {
      setHover();
    }
  }, [isSubmenuOpen, submenuCloseDelay, setHover]);
  const onMouseLeave = useCallback((_, keepHover) => {
    timeoutId.current && clearTimeout(timeoutId.current);
    if (!keepHover) dispatch({
      type: HoverIndexActionTypes.UNSET,
      index
    });
  }, [dispatch, index]);
  useEffect(() => () => clearTimeout(timeoutId.current), []);
  useEffect(() => {
    if (isHovering && isParentOpen) {
      ref.current && ref.current.focus();
    }
  }, [isHovering, isParentOpen]);
  return {
    ref,
    isHovering,
    setHover,
    onBlur,
    onMouseEnter,
    onMouseLeave
  };
};

const useMenuChange = (onChange, isOpen) => {
  const prevOpen = useRef(isOpen);
  useEffect(() => {
    if (prevOpen.current !== isOpen) safeCall(onChange, {
      open: isOpen
    });
    prevOpen.current = isOpen;
  }, [onChange, isOpen]);
};

const FocusPositions = Object.freeze({
  'INITIAL': 'initial',
  'FIRST': 'first',
  'LAST': 'last'
});
const MenuStates = Object.freeze({
  'UNMOUNTED': 0,
  'CLOSED': 1,
  'OPEN': 2
});
const MenuStateActionType = Object.freeze({
  'OPEN': 0,
  'CLOSE': 1,
  'TOGGLE': 2
});
const useMenuState = (keepMounted = true) => {
  const [menuItemFocus, setMenuItemFocus] = useState({
    position: FocusPositions.INITIAL
  });
  const [menuState, dispatch] = useReducer(menuStateReducer, MenuStates.UNMOUNTED);

  function menuStateReducer(state, {
    type
  }) {
    switch (type) {
      case MenuStateActionType.OPEN:
        return MenuStates.OPEN;

      case MenuStateActionType.CLOSE:
        if (state === MenuStates.OPEN) {
          return keepMounted ? MenuStates.CLOSED : MenuStates.UNMOUNTED;
        } else {
          return state;
        }

      case MenuStateActionType.TOGGLE:
        if (state === MenuStates.OPEN) {
          return keepMounted ? MenuStates.CLOSED : MenuStates.UNMOUNTED;
        } else {
          return MenuStates.OPEN;
        }

      default:
        throw new Error('menuStateReducer: unknown action type');
    }
  }

  return {
    isMounted: menuState !== MenuStates.UNMOUNTED,
    isOpen: menuState === MenuStates.OPEN,
    menuItemFocus,
    openMenu: useCallback((menuItemFocus = FocusPositions.INITIAL) => {
      setMenuItemFocus({
        position: menuItemFocus
      });
      dispatch({
        type: MenuStateActionType.OPEN
      });
    }, []),
    closeMenu: useCallback(() => dispatch({
      type: MenuStateActionType.CLOSE
    }), []),
    toggleMenu: useCallback((menuItemFocus = FocusPositions.INITIAL) => {
      setMenuItemFocus({
        position: menuItemFocus
      });
      dispatch({
        type: MenuStateActionType.TOGGLE
      });
    }, [])
  };
};

const useIsomorphicLayoutEffect = typeof window !== 'undefined' && typeof window.document !== 'undefined' && typeof window.document.createElement !== 'undefined' ? useLayoutEffect : useEffect;

const MenuButton = defineName( /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function MenuButton({
  className,
  styles,
  isOpen,
  disabled,
  children,
  ...restProps
}, ref) {
  const modifiers = Object.freeze({
    open: isOpen
  });
  return /*#__PURE__*/React.createElement("button", Object.assign({
    "aria-haspopup": "true",
    "aria-expanded": isOpen,
    "aria-disabled": disabled || undefined,
    disabled: disabled
  }, restProps, {
    ref: ref,
    className: bem(menuButtonClass, null, modifiers)(className),
    style: flatStyles(styles, modifiers)
  }), children);
})), 'MenuButton');
MenuButton.propTypes = { ...stylePropTypes(),
  isOpen: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func
};

const MenuList = defineName( /*#__PURE__*/React.memo(function MenuList({
  ariaLabel,
  className,
  styles,
  arrowClassName,
  arrowStyles,
  anchorPoint,
  anchorRef,
  containerRef,
  arrow,
  align,
  direction,
  position,
  overflow,
  repositionFlag,
  captureFocus = true,
  isOpen,
  isDisabled,
  menuItemFocus,
  offsetX,
  offsetY,
  children,
  onClose,
  ...restProps
}) {
  const [menuPosition, setMenuPosition] = useState({
    x: 0,
    y: 0
  });
  const [arrowPosition, setArrowPosition] = useState({});
  const [overflowData, setOverflowData] = useState();
  const [isClosing, setClosing] = useState(false);
  const [expandedDirection, setExpandedDirection] = useState(direction);
  const {
    animation,
    boundingBoxRef,
    boundingBoxPadding,
    rootMenuRef,
    rootAnchorRef,
    scrollingRef,
    anchorScrollingRef,
    reposition,
    viewScroll
  } = useContext(SettingsContext);
  const menuRef = useRef(null);
  const arrowRef = useRef(null);
  const menuItemsCount = useRef(0);
  const latestOpen = useRef(isOpen);
  const latestMenuSize = useRef({
    width: 0,
    height: 0
  });
  const latestHandlePosition = useRef(() => {});
  const descendOverflowRef = useRef(false);
  const reposFlag = useContext(MenuListContext).reposSubmenu || repositionFlag;
  const [reposSubmenu, forceReposSubmenu] = useReducer(c => c + 1, 1);
  const [{
    hoverIndex,
    openSubmenuCount
  }, dispatch] = useReducer(reducer, {
    hoverIndex: initialHoverIndex,
    openSubmenuCount: 0
  });

  function reducer({
    hoverIndex,
    openSubmenuCount
  }, action) {
    return {
      hoverIndex: hoverIndexReducer(hoverIndex, action),
      openSubmenuCount: submenuCountReducer(openSubmenuCount, action)
    };
  }

  function hoverIndexReducer(state, {
    type,
    index
  }) {
    switch (type) {
      case HoverIndexActionTypes.RESET:
        return initialHoverIndex;

      case HoverIndexActionTypes.SET:
        return index;

      case HoverIndexActionTypes.UNSET:
        return state === index ? initialHoverIndex : state;

      case HoverIndexActionTypes.DECREASE:
        {
          let i = state;
          i--;
          if (i < 0) i = menuItemsCount.current - 1;
          return i;
        }

      case HoverIndexActionTypes.INCREASE:
        {
          let i = state;
          i++;
          if (i >= menuItemsCount.current) i = 0;
          return i;
        }

      case HoverIndexActionTypes.FIRST:
        return menuItemsCount.current > 0 ? 0 : initialHoverIndex;

      case HoverIndexActionTypes.LAST:
        return menuItemsCount.current > 0 ? menuItemsCount.current - 1 : initialHoverIndex;

      default:
        return state;
    }
  }

  const menuItems = useMemo(() => {
    const {
      items,
      endIndex,
      descendOverflow
    } = cloneChildren(children);
    menuItemsCount.current = endIndex;
    descendOverflowRef.current = descendOverflow;
    return items;
  }, [children]);

  const handleKeyDown = e => {
    let handled = false;

    switch (e.key) {
      case Keys.HOME:
        dispatch({
          type: HoverIndexActionTypes.FIRST
        });
        handled = true;
        break;

      case Keys.END:
        dispatch({
          type: HoverIndexActionTypes.LAST
        });
        handled = true;
        break;

      case Keys.UP:
        dispatch({
          type: HoverIndexActionTypes.DECREASE
        });
        handled = true;
        break;

      case Keys.DOWN:
        dispatch({
          type: HoverIndexActionTypes.INCREASE
        });
        handled = true;
        break;

      case Keys.SPACE:
        if (e.target && e.target.className.includes(menuClass)) {
          e.preventDefault();
        }

        break;
    }

    if (handled) {
      e.preventDefault();
      e.stopPropagation();
    }
  };

  const handleAnimationEnd = () => {
    if (isClosing) {
      setClosing(false);
      setOverflowData();
    }
  };

  const positionHelpers = useCallback(() => {
    const menuRect = menuRef.current.getBoundingClientRect();
    const containerRect = containerRef.current.getBoundingClientRect();
    const boundingRect = scrollingRef.current === window ? {
      left: 0,
      top: 0,
      right: document.documentElement.clientWidth,
      bottom: window.innerHeight
    } : scrollingRef.current.getBoundingClientRect();
    const padding = parsePadding(boundingBoxPadding);

    const getLeftOverflow = x => x + containerRect.left - boundingRect.left - padding.left;

    const getRightOverflow = x => x + containerRect.left + menuRect.width - boundingRect.right + padding.right;

    const getTopOverflow = y => y + containerRect.top - boundingRect.top - padding.top;

    const getBottomOverflow = y => y + containerRect.top + menuRect.height - boundingRect.bottom + padding.bottom;

    const confineHorizontally = x => {
      let leftOverflow = getLeftOverflow(x);

      if (leftOverflow < 0) {
        x -= leftOverflow;
      } else {
        const rightOverflow = getRightOverflow(x);

        if (rightOverflow > 0) {
          x -= rightOverflow;
          leftOverflow = getLeftOverflow(x);
          if (leftOverflow < 0) x -= leftOverflow;
        }
      }

      return x;
    };

    const confineVertically = y => {
      let topOverflow = getTopOverflow(y);

      if (topOverflow < 0) {
        y -= topOverflow;
      } else {
        const bottomOverflow = getBottomOverflow(y);

        if (bottomOverflow > 0) {
          y -= bottomOverflow;
          topOverflow = getTopOverflow(y);
          if (topOverflow < 0) y -= topOverflow;
        }
      }

      return y;
    };

    return {
      menuRect,
      containerRect,
      getLeftOverflow,
      getRightOverflow,
      getTopOverflow,
      getBottomOverflow,
      confineHorizontally,
      confineVertically
    };
  }, [containerRef, scrollingRef, boundingBoxPadding]);
  const placeArrowX = useCallback((menuX, anchorRect, containerRect, menuRect) => {
    let x = anchorRect.left - containerRect.left - menuX + anchorRect.width / 2;
    const offset = arrowRef.current.offsetWidth * 1.25;
    x = Math.max(offset, x);
    x = Math.min(x, menuRect.width - offset);
    return x;
  }, []);
  const placeArrowY = useCallback((menuY, anchorRect, containerRect, menuRect) => {
    let y = anchorRect.top - containerRect.top - menuY + anchorRect.height / 2;
    const offset = arrowRef.current.offsetHeight * 1.25;
    y = Math.max(offset, y);
    y = Math.min(y, menuRect.height - offset);
    return y;
  }, []);
  const placeLeftorRight = useCallback(({
    anchorRect,
    containerRect,
    menuRect
  }, {
    placeLeftorRightY,
    placeLeftX,
    placeRightX
  }, {
    getLeftOverflow,
    getRightOverflow,
    confineHorizontally,
    confineVertically
  }) => {
    let computedDirection = direction;
    let y = placeLeftorRightY;

    if (position !== 'initial') {
      y = confineVertically(y);

      if (position === 'anchor') {
        y = Math.min(y, anchorRect.bottom - containerRect.top);
        y = Math.max(y, anchorRect.top - containerRect.top - menuRect.height);
      }
    }

    let x, leftOverflow, rightOverflow;

    if (computedDirection === 'left') {
      x = placeLeftX;

      if (position !== 'initial') {
        leftOverflow = getLeftOverflow(x);

        if (leftOverflow < 0) {
          rightOverflow = getRightOverflow(placeRightX);

          if (rightOverflow <= 0 || -leftOverflow > rightOverflow) {
            x = placeRightX;
            computedDirection = 'right';
          }
        }
      }
    } else {
      x = placeRightX;

      if (position !== 'initial') {
        rightOverflow = getRightOverflow(x);

        if (rightOverflow > 0) {
          leftOverflow = getLeftOverflow(placeLeftX);

          if (leftOverflow >= 0 || -leftOverflow < rightOverflow) {
            x = placeLeftX;
            computedDirection = 'left';
          }
        }
      }
    }

    if (position === 'auto') x = confineHorizontally(x);
    const arrowY = arrow ? placeArrowY(y, anchorRect, containerRect, menuRect) : undefined;
    return {
      arrowY,
      x,
      y,
      computedDirection
    };
  }, [placeArrowY, arrow, direction, position]);
  const placeToporBottom = useCallback(({
    anchorRect,
    containerRect,
    menuRect
  }, {
    placeToporBottomX,
    placeTopY,
    placeBottomY
  }, {
    getTopOverflow,
    getBottomOverflow,
    confineHorizontally,
    confineVertically
  }) => {
    let computedDirection = direction === 'top' ? 'top' : 'bottom';
    let x = placeToporBottomX;

    if (position !== 'initial') {
      x = confineHorizontally(x);

      if (position === 'anchor') {
        x = Math.min(x, anchorRect.right - containerRect.left);
        x = Math.max(x, anchorRect.left - containerRect.left - menuRect.width);
      }
    }

    let y, topOverflow, bottomOverflow;

    if (computedDirection === 'top') {
      y = placeTopY;

      if (position !== 'initial') {
        topOverflow = getTopOverflow(y);

        if (topOverflow < 0) {
          bottomOverflow = getBottomOverflow(placeBottomY);

          if (bottomOverflow <= 0 || -topOverflow > bottomOverflow) {
            y = placeBottomY;
            computedDirection = 'bottom';
          }
        }
      }
    } else {
      y = placeBottomY;

      if (position !== 'initial') {
        bottomOverflow = getBottomOverflow(y);

        if (bottomOverflow > 0) {
          topOverflow = getTopOverflow(placeTopY);

          if (topOverflow >= 0 || -topOverflow < bottomOverflow) {
            y = placeTopY;
            computedDirection = 'top';
          }
        }
      }
    }

    if (position === 'auto') y = confineVertically(y);
    const arrowX = arrow ? placeArrowX(x, anchorRect, containerRect, menuRect) : undefined;
    return {
      arrowX,
      x,
      y,
      computedDirection
    };
  }, [placeArrowX, arrow, direction, position]);
  const positionMenu = useCallback((positionHelpers, anchorRef) => {
    const {
      menuRect,
      containerRect,
      ...helpers
    } = positionHelpers;
    let horizontalOffset = offsetX;
    let verticalOffset = offsetY;

    if (arrow) {
      if (direction === 'left' || direction === 'right') {
        horizontalOffset += arrowRef.current.offsetWidth;
      } else {
        verticalOffset += arrowRef.current.offsetHeight;
      }
    }

    const anchorRect = anchorRef.current.getBoundingClientRect();
    const placeLeftX = anchorRect.left - containerRect.left - menuRect.width - horizontalOffset;
    const placeRightX = anchorRect.right - containerRect.left + horizontalOffset;
    const placeTopY = anchorRect.top - containerRect.top - menuRect.height - verticalOffset;
    const placeBottomY = anchorRect.bottom - containerRect.top + verticalOffset;
    let placeToporBottomX, placeLeftorRightY;

    if (align === 'end') {
      placeToporBottomX = anchorRect.right - containerRect.left - menuRect.width;
      placeLeftorRightY = anchorRect.bottom - containerRect.top - menuRect.height;
    } else if (align === 'center') {
      placeToporBottomX = anchorRect.left - containerRect.left - (menuRect.width - anchorRect.width) / 2;
      placeLeftorRightY = anchorRect.top - containerRect.top - (menuRect.height - anchorRect.height) / 2;
    } else {
      placeToporBottomX = anchorRect.left - containerRect.left;
      placeLeftorRightY = anchorRect.top - containerRect.top;
    }

    placeToporBottomX += horizontalOffset;
    placeLeftorRightY += verticalOffset;
    const rects = {
      anchorRect,
      containerRect,
      menuRect
    };
    const placements = {
      placeLeftX,
      placeRightX,
      placeLeftorRightY,
      placeTopY,
      placeBottomY,
      placeToporBottomX
    };

    switch (direction) {
      case 'left':
      case 'right':
        return placeLeftorRight(rects, placements, helpers);

      case 'top':
      case 'bottom':
      default:
        return placeToporBottom(rects, placements, helpers);
    }
  }, [arrow, align, direction, offsetX, offsetY, placeLeftorRight, placeToporBottom]);
  const positionContextMenu = useCallback((positionHelpers, anchorPoint) => {
    const {
      menuRect,
      containerRect,
      getLeftOverflow,
      getRightOverflow,
      getTopOverflow,
      getBottomOverflow,
      confineHorizontally,
      confineVertically
    } = positionHelpers;
    let x, y;
    x = anchorPoint.x - containerRect.left;
    y = anchorPoint.y - containerRect.top;
    const rightOverflow = getRightOverflow(x);

    if (rightOverflow > 0) {
      const adjustedX = x - menuRect.width;
      const leftOverflow = getLeftOverflow(adjustedX);

      if (leftOverflow >= 0 || -leftOverflow < rightOverflow) {
        x = adjustedX;
      }

      x = confineHorizontally(x);
    }

    let computedDirection = 'bottom';
    const bottomOverflow = getBottomOverflow(y);

    if (bottomOverflow > 0) {
      const adjustedY = y - menuRect.height;
      const topOverflow = getTopOverflow(adjustedY);

      if (topOverflow >= 0 || -topOverflow < bottomOverflow) {
        y = adjustedY;
        computedDirection = 'top';
      }

      y = confineVertically(y);
    }

    return {
      x,
      y,
      computedDirection
    };
  }, []);
  const handlePosition = useCallback(() => {
    if (!containerRef.current) {
      if (!isProd) throw new Error('[react-menu] Menu cannot be positioned properly as container ref is null. If you initialise isOpen prop to true for ControlledMenu, please see this link for a solution: https://github.com/szhsin/react-menu/issues/2#issuecomment-719166062');
      return;
    }

    if (!scrollingRef.current) {
      scrollingRef.current = boundingBoxRef ? boundingBoxRef.current : getScrollAncestor(rootMenuRef.current);
    }

    const helpers = positionHelpers();
    const {
      menuRect
    } = helpers;
    let results = {
      computedDirection: 'bottom'
    };

    if (anchorPoint) {
      results = positionContextMenu(helpers, anchorPoint);
    } else if (anchorRef) {
      results = positionMenu(helpers, anchorRef);
    }

    let {
      arrowX,
      arrowY,
      x,
      y,
      computedDirection
    } = results;
    let menuHeight = menuRect.height;

    if (overflow !== 'visible') {
      const {
        getTopOverflow,
        getBottomOverflow
      } = helpers;
      let height, overflowAmt;
      const prevHeight = latestMenuSize.current.height;
      const bottomOverflow = getBottomOverflow(y);

      if (bottomOverflow > 0 || floatEqual(bottomOverflow, 0) && floatEqual(menuHeight, prevHeight)) {
        height = menuHeight - bottomOverflow;
        overflowAmt = bottomOverflow;
      } else {
        const topOverflow = getTopOverflow(y);

        if (topOverflow < 0 || floatEqual(topOverflow, 0) && floatEqual(menuHeight, prevHeight)) {
          height = menuHeight + topOverflow;
          overflowAmt = 0 - topOverflow;
          if (height >= 0) y -= topOverflow;
        }
      }

      if (height >= 0) {
        menuHeight = height;
        setOverflowData({
          height,
          overflowAmt
        });
      } else {
        setOverflowData();
      }
    }

    if (arrow) setArrowPosition({
      x: arrowX,
      y: arrowY
    });
    setMenuPosition({
      x,
      y
    });
    setExpandedDirection(computedDirection);
    latestMenuSize.current = {
      width: menuRect.width,
      height: menuHeight
    };
  }, [arrow, anchorPoint, anchorRef, containerRef, boundingBoxRef, rootMenuRef, scrollingRef, overflow, positionHelpers, positionMenu, positionContextMenu]);
  useIsomorphicLayoutEffect(() => {
    if (isOpen) {
      handlePosition();
      if (latestOpen.current) forceReposSubmenu();
    }

    latestHandlePosition.current = handlePosition;
  }, [isOpen, handlePosition, reposFlag]);
  useIsomorphicLayoutEffect(() => {
    if (overflowData && !descendOverflowRef.current) menuRef.current.scrollTop = 0;
  }, [overflowData]);
  useIsomorphicLayoutEffect(() => {
    if (animation) {
      if (isOpen) {
        setClosing(false);
      } else if (isOpen !== latestOpen.current) {
        setClosing(true);
      }
    }

    latestOpen.current = isOpen;
  }, [animation, isOpen]);
  useEffect(() => {
    if (!isOpen) return;

    if (!anchorScrollingRef.current && rootAnchorRef && rootAnchorRef.current.tagName) {
      anchorScrollingRef.current = getScrollAncestor(rootAnchorRef.current);
    }

    const scrollCurrent = scrollingRef.current;
    const menuScroll = scrollCurrent && scrollCurrent.addEventListener ? scrollCurrent : window;
    const anchorScroll = anchorScrollingRef.current || menuScroll;
    let scroll = viewScroll;
    if (anchorScroll !== menuScroll && scroll === 'initial') scroll = 'auto';
    if (scroll === 'initial') return;
    if (scroll === 'auto' && overflow !== 'visible') scroll = 'close';

    const handleScroll = () => {
      if (scroll === 'auto') {
        batchedUpdates(handlePosition);
      } else {
        safeCall(onClose, {
          reason: CloseReason.SCROLL
        });
      }
    };

    const scrollObservers = anchorScroll !== menuScroll && viewScroll !== 'initial' ? [anchorScroll, menuScroll] : [anchorScroll];
    scrollObservers.forEach(o => o.addEventListener('scroll', handleScroll));
    return () => scrollObservers.forEach(o => o.removeEventListener('scroll', handleScroll));
  }, [rootAnchorRef, anchorScrollingRef, scrollingRef, isOpen, overflow, onClose, viewScroll, handlePosition]);
  useEffect(() => {
    if (typeof ResizeObserver !== 'function' || reposition === 'initial') return;
    const resizeObserver = new ResizeObserver(([entry]) => {
      const {
        borderBoxSize,
        target
      } = entry;
      let width, height;

      if (borderBoxSize) {
        const {
          inlineSize,
          blockSize
        } = borderBoxSize[0] || borderBoxSize;
        width = inlineSize;
        height = blockSize;
      } else {
        const borderRect = target.getBoundingClientRect();
        width = borderRect.width;
        height = borderRect.height;
      }

      if (width === 0 || height === 0) return;
      if (floatEqual(width, latestMenuSize.current.width, 1) && floatEqual(height, latestMenuSize.current.height, 1)) return;
      batchedUpdates(() => {
        latestHandlePosition.current();
        forceReposSubmenu();
      });
    });
    const observeTarget = menuRef.current;
    resizeObserver.observe(observeTarget, {
      box: 'border-box'
    });
    return () => resizeObserver.unobserve(observeTarget);
  }, [reposition]);
  useEffect(() => {
    if (!isOpen) {
      dispatch({
        type: HoverIndexActionTypes.RESET
      });
      if (!animation) setOverflowData();
    }

    const id = setTimeout(() => {
      if (!isOpen || !menuRef.current || menuRef.current.contains(document.activeElement)) return;
      if (captureFocus) menuRef.current.focus();

      if (menuItemFocus.position === FocusPositions.FIRST) {
        dispatch({
          type: HoverIndexActionTypes.FIRST
        });
      } else if (menuItemFocus.position === FocusPositions.LAST) {
        dispatch({
          type: HoverIndexActionTypes.LAST
        });
      }
    }, animation ? 170 : 100);
    return () => clearTimeout(id);
  }, [animation, captureFocus, isOpen, menuItemFocus]);
  const isSubmenuOpen = openSubmenuCount > 0;
  const itemContext = useMemo(() => ({
    isParentOpen: isOpen,
    hoverIndex,
    isSubmenuOpen,
    dispatch
  }), [isOpen, hoverIndex, isSubmenuOpen]);
  let maxHeight, overflowAmt;

  if (overflowData) {
    descendOverflowRef.current ? overflowAmt = overflowData.overflowAmt : maxHeight = overflowData.height;
  }

  const listContext = useMemo(() => ({
    reposSubmenu,
    overflow,
    overflowAmt
  }), [reposSubmenu, overflow, overflowAmt]);
  const overflowStyles = maxHeight >= 0 ? {
    maxHeight,
    overflow
  } : undefined;
  const modifiers = {
    open: isOpen,
    closing: isClosing,
    animation,
    dir: animation && expandedDirection
  };
  const userModifiers = Object.freeze({ ...modifiers,
    dir: expandedDirection
  });
  const arrowModifiers = Object.freeze({
    dir: expandedDirection
  });
  const handlers = attachHandlerProps({
    onKeyDown: handleKeyDown,
    onAnimationEnd: handleAnimationEnd
  }, restProps);
  return /*#__PURE__*/React.createElement("ul", Object.assign({
    role: "menu",
    tabIndex: "-1",
    "aria-disabled": isDisabled || undefined,
    "aria-label": ariaLabel
  }, restProps, handlers, {
    ref: menuRef,
    className: bem(menuClass, null, modifiers)(className, userModifiers),
    style: { ...flatStyles(styles, userModifiers),
      ...overflowStyles,
      left: `${menuPosition.x}px`,
      top: `${menuPosition.y}px`
    }
  }), arrow && /*#__PURE__*/React.createElement("div", {
    className: bem(menuClass, menuArrowClass, arrowModifiers)(arrowClassName),
    style: { ...flatStyles(arrowStyles, arrowModifiers),
      left: arrowPosition.x && `${arrowPosition.x}px`,
      top: arrowPosition.y && `${arrowPosition.y}px`
    },
    ref: arrowRef
  }), /*#__PURE__*/React.createElement(MenuListContext.Provider, {
    value: listContext
  }, /*#__PURE__*/React.createElement(MenuListItemContext.Provider, {
    value: itemContext
  }, menuItems)));
}), 'MenuList');

function submenuCountReducer(state, {
  type
}) {
  switch (type) {
    case SubmenuActionTypes.OPEN:
      return state + 1;

    case SubmenuActionTypes.CLOSE:
      return Math.max(state - 1, 0);

    default:
      return state;
  }
}

const useMenuList = (menuListProps, {
  id,
  animation,
  boundingBoxRef,
  boundingBoxPadding,
  debugging,
  reposition,
  submenuOpenDelay,
  submenuCloseDelay,
  viewScroll,
  portal,
  theming,
  isMounted,
  onClick,
  onClose,
  skipClick
}) => {
  const containerRef = useRef(null);
  const scrollingRef = useRef(null);
  const anchorScrollingRef = useRef(null);
  const {
    anchorRef,
    isOpen
  } = menuListProps;
  const settings = useMemo(() => ({
    animation,
    boundingBoxRef,
    boundingBoxPadding,
    rootMenuRef: containerRef,
    rootAnchorRef: anchorRef,
    scrollingRef,
    anchorScrollingRef,
    reposition,
    viewScroll
  }), [animation, anchorRef, boundingBoxRef, boundingBoxPadding, reposition, viewScroll]);
  const itemSettings = useMemo(() => ({
    debugging,
    submenuOpenDelay,
    submenuCloseDelay
  }), [debugging, submenuOpenDelay, submenuCloseDelay]);
  const eventHandlers = useMemo(() => ({
    handleClick(event, isStopPropagation, isCheckorRadio) {
      if (!isStopPropagation) safeCall(onClick, event);
      let keepOpen = event.keepOpen;

      if (keepOpen === undefined) {
        keepOpen = isCheckorRadio && event.key === Keys.SPACE;
      }

      if (!keepOpen) {
        safeCall(onClose, {
          value: event.value,
          key: event.key,
          reason: CloseReason.CLICK
        });
      }
    },

    handleClose(key) {
      safeCall(onClose, {
        key,
        reason: CloseReason.CLICK
      });
    }

  }), [onClick, onClose]);

  const handleKeyDown = ({
    key
  }) => {
    switch (key) {
      case Keys.ESC:
        safeCall(onClose, {
          key,
          reason: CloseReason.CANCEL
        });
        break;
    }
  };

  const handleBlur = e => {
    if (isOpen && !e.currentTarget.contains(e.relatedTarget || document.activeElement) && !debugging) {
      safeCall(onClose, {
        reason: CloseReason.BLUR
      });

      if (skipClick) {
        skipClick.current = true;
        setTimeout(() => skipClick.current = false, 300);
      }
    }
  };

  const menuList = /*#__PURE__*/React.createElement("div", {
    id: id,
    className: bem(menuContainerClass, null, {
      theme: theming
    })(),
    ref: containerRef,
    onKeyDown: handleKeyDown,
    onBlur: handleBlur
  }, isMounted && /*#__PURE__*/React.createElement(SettingsContext.Provider, {
    value: settings
  }, /*#__PURE__*/React.createElement(ItemSettingsContext.Provider, {
    value: itemSettings
  }, /*#__PURE__*/React.createElement(EventHandlersContext.Provider, {
    value: eventHandlers
  }, /*#__PURE__*/React.createElement(MenuList, Object.assign({}, menuListProps, {
    containerRef: containerRef,
    onClose: onClose
  }))))));

  if (portal) {
    return /*#__PURE__*/ReactDOM.createPortal(menuList, document.body);
  } else {
    return menuList;
  }
};

const Menu = /*#__PURE__*/React.memo(function Menu({
  'aria-label': ariaLabel,
  id,
  animation,
  boundingBoxRef,
  boundingBoxPadding,
  captureFocus: _,
  debugging,
  reposition,
  viewScroll,
  keepMounted,
  menuButton,
  portal,
  submenuOpenDelay,
  submenuCloseDelay,
  theming,
  onClick,
  onChange,
  ...restProps
}) {
  const {
    isMounted,
    isOpen,
    menuItemFocus,
    openMenu,
    closeMenu
  } = useMenuState(keepMounted);
  const skipClick = useRef(false);
  const buttonRef = useRef(null);
  const button = safeCall(menuButton, {
    open: isOpen
  });
  if (!button) throw new Error('Menu requires a menuButton prop.');
  const handleClose = useCallback(e => {
    closeMenu();
    if (e.key) buttonRef.current.focus();
  }, [closeMenu]);
  const handleClick = useCallback(e => {
    if (skipClick.current) return;
    openMenu(e.detail === 0 ? FocusPositions.FIRST : FocusPositions.INITIAL);
  }, [openMenu]);
  const handleKeyDown = useCallback(e => {
    let handled = false;

    switch (e.key) {
      case Keys.UP:
        openMenu(FocusPositions.LAST);
        handled = true;
        break;

      case Keys.DOWN:
        openMenu(FocusPositions.FIRST);
        handled = true;
        break;
    }

    if (handled) e.preventDefault();
  }, [openMenu]);
  const renderButton = useMemo(() => {
    const buttonProps = {
      ref: buttonRef,
      ...attachHandlerProps({
        onClick: handleClick,
        onKeyDown: handleKeyDown
      }, button.props)
    };

    if (getName(button.type) === 'MenuButton') {
      buttonProps.isOpen = isOpen;
    }

    return /*#__PURE__*/React.cloneElement(button, buttonProps);
  }, [button, isOpen, handleClick, handleKeyDown]);
  const menuList = useMenuList({ ...restProps,
    ariaLabel: ariaLabel || (typeof button.props.children === 'string' ? button.props.children : 'Menu'),
    anchorRef: buttonRef,
    isOpen,
    menuItemFocus
  }, {
    id,
    animation,
    boundingBoxRef,
    boundingBoxPadding,
    debugging,
    reposition,
    submenuOpenDelay,
    submenuCloseDelay,
    viewScroll,
    portal,
    theming,
    isMounted,
    onClick,
    onClose: handleClose,
    skipClick
  });
  useMenuChange(onChange, isOpen);
  return /*#__PURE__*/React.createElement(React.Fragment, null, renderButton, menuList);
});
Menu.propTypes = { ...menuPropTypesBase,
  keepMounted: PropTypes.bool,
  menuButton: PropTypes.oneOfType([PropTypes.element, PropTypes.func]).isRequired,
  onChange: PropTypes.func
};
Menu.defaultProps = { ...menuDefaultPropsBase,
  keepMounted: true
};

const ControlledMenu = /*#__PURE__*/React.memo(function ControlledMenu({
  'aria-label': ariaLabel,
  id,
  animation,
  boundingBoxRef,
  boundingBoxPadding,
  debugging,
  reposition,
  submenuOpenDelay,
  submenuCloseDelay,
  viewScroll,
  portal,
  theming,
  isMounted,
  onClick,
  onClose,
  ...restProps
}) {
  return useMenuList({ ...restProps,
    ariaLabel: ariaLabel || 'Menu'
  }, {
    id,
    animation,
    boundingBoxRef,
    boundingBoxPadding,
    debugging,
    reposition,
    submenuOpenDelay,
    submenuCloseDelay,
    viewScroll,
    portal,
    theming,
    isMounted,
    onClick,
    onClose
  });
});
ControlledMenu.propTypes = { ...menuPropTypesBase,
  anchorPoint: PropTypes.exact({
    x: PropTypes.number,
    y: PropTypes.number
  }),
  anchorRef: PropTypes.object,
  captureFocus: PropTypes.bool,
  isOpen: PropTypes.bool,
  isMounted: PropTypes.bool,
  menuItemFocus: PropTypes.exact({
    position: PropTypes.string
  }),
  onClose: PropTypes.func
};
ControlledMenu.defaultProps = { ...menuDefaultPropsBase,
  captureFocus: true,
  isOpen: false,
  isMounted: true,
  menuItemFocus: {
    position: FocusPositions.INITIAL
  }
};

const SubMenu = defineName( /*#__PURE__*/React.memo(function SubMenu({
  'aria-label': ariaLabel,
  itemClassName,
  itemStyles,
  disabled,
  keepMounted,
  label,
  index,
  onChange,
  captureFocus: _1,
  repositionFlag: _2,
  ...restProps
}) {
  const {
    isMounted,
    isOpen,
    menuItemFocus,
    openMenu,
    closeMenu
  } = useMenuState(keepMounted);
  const {
    isParentOpen,
    hoverIndex,
    isSubmenuOpen,
    dispatch
  } = useContext(MenuListItemContext);
  const {
    debugging,
    submenuOpenDelay,
    submenuCloseDelay
  } = useContext(ItemSettingsContext);
  const isHovering = hoverIndex === index;
  const isDisabled = Boolean(disabled);
  const {
    isActive,
    onKeyUp,
    ...activeStateHandlers
  } = useActiveState(isHovering, isDisabled, Keys.RIGHT);
  const containerRef = useRef(null);
  const itemRef = useRef(null);
  const timeoutId = useRef();
  const handleClose = useCallback(() => {
    closeMenu();
    itemRef.current.focus();
  }, [closeMenu]);

  const delayOpen = delay => {
    dispatch({
      type: HoverIndexActionTypes.SET,
      index
    });
    timeoutId.current = setTimeout(openMenu, Math.max(delay, 0));
  };

  const handleMouseEnter = () => {
    if (isDisabled || isOpen) return;

    if (isSubmenuOpen) {
      timeoutId.current = setTimeout(() => delayOpen(submenuOpenDelay - submenuCloseDelay), submenuCloseDelay);
    } else {
      delayOpen(submenuOpenDelay);
    }
  };

  const handleMouseLeave = () => {
    clearTimeout(timeoutId.current);

    if (!isOpen) {
      dispatch({
        type: HoverIndexActionTypes.UNSET,
        index
      });
    }
  };

  const handleClick = () => {
    if (isDisabled) return;
    clearTimeout(timeoutId.current);
    openMenu();
  };

  const handleKeyDown = e => {
    let handled = false;

    switch (e.key) {
      case Keys.LEFT:
        if (isOpen) {
          handleClose();
          handled = true;
        }

        break;

      case Keys.RIGHT:
        if (!isOpen) handled = true;
        break;
    }

    if (handled) {
      e.preventDefault();
      e.stopPropagation();
    }
  };

  const handleKeyUp = e => {
    if (!isActive) return;
    onKeyUp(e);

    switch (e.key) {
      case Keys.SPACE:
      case Keys.ENTER:
      case Keys.RIGHT:
        openMenu(FocusPositions.FIRST);
        break;
    }
  };

  const handleBlur = e => {
    if (debugging) return;
    const relatedTarget = e.relatedTarget || document.activeElement;

    if (!e.currentTarget.contains(relatedTarget)) {
      closeMenu();
      dispatch({
        type: HoverIndexActionTypes.UNSET,
        index
      });
    }
  };

  useEffect(() => () => clearTimeout(timeoutId.current), []);
  useEffect(() => {
    if (isHovering && isParentOpen) {
      itemRef.current.focus();
    } else {
      closeMenu();
    }
  }, [isHovering, isParentOpen, closeMenu]);
  useEffect(() => {
    dispatch({
      type: isOpen ? SubmenuActionTypes.OPEN : SubmenuActionTypes.CLOSE
    });
  }, [dispatch, isOpen]);
  useMenuChange(onChange, isOpen);
  const modifiers = Object.freeze({
    open: isOpen,
    hover: isHovering,
    active: isActive,
    disabled: isDisabled
  });
  return /*#__PURE__*/React.createElement("li", {
    className: bem(menuClass, subMenuClass)(),
    role: "presentation",
    ref: containerRef,
    onKeyDown: handleKeyDown,
    onBlur: handleBlur
  }, /*#__PURE__*/React.createElement("div", Object.assign({
    className: bem(menuClass, menuItemClass, modifiers)(itemClassName),
    style: flatStyles(itemStyles, modifiers),
    role: "menuitem",
    "aria-haspopup": "true",
    "aria-expanded": isOpen,
    "aria-disabled": isDisabled || undefined,
    tabIndex: isHovering && !isOpen ? 0 : -1,
    ref: itemRef,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onMouseDown: () => !isHovering && dispatch({
      type: HoverIndexActionTypes.SET,
      index
    }),
    onClick: handleClick,
    onKeyUp: handleKeyUp
  }, activeStateHandlers), safeCall(label, modifiers)), isMounted && /*#__PURE__*/React.createElement(MenuList, Object.assign({}, restProps, {
    ariaLabel: ariaLabel || (typeof label === 'string' ? label : 'Submenu'),
    anchorRef: itemRef,
    containerRef: containerRef,
    isOpen: isOpen,
    isDisabled: isDisabled,
    menuItemFocus: menuItemFocus
  })));
}), 'SubMenu');
SubMenu.propTypes = { ...sharedMenuPropTypes,
  ...stylePropTypes('item'),
  disabled: PropTypes.bool,
  keepMounted: PropTypes.bool,
  label: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  onChange: PropTypes.func
};
SubMenu.defaultProps = { ...sharedMenuDefaultProp,
  direction: 'right',
  keepMounted: true
};

const MenuItem = defineName( /*#__PURE__*/React.memo(function MenuItem({
  className,
  styles,
  value,
  href,
  type,
  checked,
  disabled,
  index,
  children,
  onClick,
  ...restProps
}) {
  const isDisabled = Boolean(disabled);
  const {
    ref,
    isHovering,
    setHover,
    onBlur,
    onMouseEnter,
    onMouseLeave
  } = useItemState(isDisabled, index);
  const eventHandlers = useContext(EventHandlersContext);
  const radioGroup = useContext(RadioGroupContext);
  const {
    isActive,
    onKeyUp,
    onBlur: activeStateBlur,
    ...activeStateHandlers
  } = useActiveState(isHovering, isDisabled);
  const isRadio = type === 'radio';
  const isCheckBox = type === 'checkbox';
  const isAnchor = Boolean(href) && !isDisabled && !isRadio && !isCheckBox;
  const isChecked = isRadio ? radioGroup.value === value : isCheckBox ? Boolean(checked) : false;

  const handleClick = key => {
    if (isDisabled) return;
    let isStopPropagation = false;
    const event = {
      value,
      key
    };

    if (isRadio) {
      event.name = radioGroup.name;
      safeCall(radioGroup.onChange, event);
    }

    event.checked = isCheckBox ? !isChecked : false;
    isStopPropagation = safeCall(onClick, event) === false;
    eventHandlers.handleClick(event, isStopPropagation, isCheckBox || isRadio);
  };

  const handleKeyUp = e => {
    if (!isActive) return;
    onKeyUp(e);

    switch (e.key) {
      case Keys.SPACE:
      case Keys.ENTER:
        if (isAnchor) {
          ref.current.click();
        } else {
          handleClick(e.key);
        }

        break;
    }
  };

  const handleBlur = e => {
    activeStateBlur(e);
    onBlur(e);
  };

  const modifiers = Object.freeze({
    type,
    disabled: isDisabled,
    hover: isHovering,
    active: isActive,
    checked: isChecked,
    anchor: isAnchor
  });
  const handlers = attachHandlerProps({ ...activeStateHandlers,
    onMouseEnter,
    onMouseLeave,
    onMouseDown: setHover,
    onKeyUp: handleKeyUp,
    onBlur: handleBlur,
    onClick: () => handleClick()
  }, restProps);
  const menuItemProps = {
    role: isRadio ? 'menuitemradio' : isCheckBox ? 'menuitemcheckbox' : 'menuitem',
    'aria-checked': isRadio || isCheckBox ? isChecked : undefined,
    'aria-disabled': isDisabled || undefined,
    tabIndex: isHovering ? 0 : -1,
    ...restProps,
    ...handlers,
    ref,
    className: bem(menuClass, menuItemClass, modifiers)(className),
    style: flatStyles(styles, modifiers)
  };
  const renderChildren = safeCall(children, modifiers);

  if (isAnchor) {
    return /*#__PURE__*/React.createElement("li", {
      role: "presentation"
    }, /*#__PURE__*/React.createElement("a", Object.assign({}, menuItemProps, {
      href: href
    }), renderChildren));
  } else {
    return /*#__PURE__*/React.createElement("li", menuItemProps, renderChildren);
  }
}), 'MenuItem');
MenuItem.propTypes = { ...stylePropTypes(),
  value: PropTypes.any,
  href: PropTypes.string,
  type: PropTypes.oneOf(['checkbox', 'radio']),
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  onClick: PropTypes.func
};

const FocusableItem = defineName( /*#__PURE__*/React.memo(function FocusableItem({
  className,
  styles,
  disabled,
  index,
  children,
  ...restProps
}) {
  const isDisabled = Boolean(disabled);
  const {
    ref,
    isHovering,
    setHover,
    onBlur,
    onMouseEnter,
    onMouseLeave
  } = useItemState(isDisabled, index);
  const {
    handleClose
  } = useContext(EventHandlersContext);
  const baseParams = {
    disabled: isDisabled,
    hover: isHovering
  };
  const modifiers = Object.freeze({ ...baseParams,
    focusable: true
  });
  const renderChildren = safeCall(children, { ...baseParams,
    ref,
    closeMenu: handleClose
  });
  const handlers = attachHandlerProps({
    onMouseEnter,
    onMouseLeave: e => onMouseLeave(e, true),
    onFocus: setHover,
    onBlur
  }, restProps);
  return /*#__PURE__*/React.createElement("li", Object.assign({
    "aria-disabled": isDisabled || undefined,
    role: "menuitem",
    tabIndex: "-1"
  }, restProps, handlers, {
    className: bem(menuClass, menuItemClass, modifiers)(className),
    style: flatStyles(styles, modifiers)
  }), renderChildren);
}), 'FocusableItem');
FocusableItem.propTypes = { ...stylePropTypes(),
  disabled: PropTypes.bool,
  children: PropTypes.func.isRequired
};

const MenuDivider = defineName( /*#__PURE__*/React.memo(function MenuDivider({
  className,
  styles,
  ...restProps
}) {
  return /*#__PURE__*/React.createElement("li", Object.assign({
    role: "separator"
  }, restProps, {
    className: bem(menuClass, menuDividerClass)(className),
    style: flatStyles(styles)
  }));
}), 'MenuDivider');
MenuDivider.propTypes = { ...stylePropTypes()
};

const MenuHeader = defineName( /*#__PURE__*/React.memo(function MenuHeader({
  className,
  styles,
  ...restProps
}) {
  return /*#__PURE__*/React.createElement("li", Object.assign({
    role: "presentation"
  }, restProps, {
    className: bem(menuClass, menuHeaderClass)(className),
    style: flatStyles(styles)
  }));
}), 'MenuHeader');
MenuHeader.propTypes = { ...stylePropTypes()
};

const MenuGroup = defineName( /*#__PURE__*/React.memo(function MenuGroup({
  className,
  styles,
  takeOverflow,
  ...restProps
}) {
  const ref = useRef(null);
  const [overflowStyles, setOverflowStyles] = useState();
  const {
    overflow,
    overflowAmt
  } = useContext(MenuListContext);
  useIsomorphicLayoutEffect(() => {
    let maxHeight;

    if (takeOverflow && overflowAmt >= 0) {
      maxHeight = ref.current.getBoundingClientRect().height - overflowAmt;
      if (maxHeight < 0) maxHeight = 0;
    }

    setOverflowStyles(maxHeight >= 0 ? {
      maxHeight,
      overflow
    } : undefined);
  }, [takeOverflow, overflow, overflowAmt]);
  useIsomorphicLayoutEffect(() => {
    if (overflowStyles) ref.current.scrollTop = 0;
  }, [overflowStyles]);
  return /*#__PURE__*/React.createElement("div", Object.assign({
    ref: ref
  }, restProps, {
    className: bem(menuClass, menuGroupClass)(className),
    style: { ...flatStyles(styles),
      ...overflowStyles
    }
  }));
}), 'MenuGroup');
MenuGroup.propTypes = { ...stylePropTypes(),
  takeOverflow: PropTypes.bool
};

const MenuRadioGroup = defineName( /*#__PURE__*/React.memo(function MenuRadioGroup({
  'aria-label': ariaLabel,
  className,
  styles,
  name,
  value,
  children,
  onChange,
  ...restProps
}) {
  const contextValue = useMemo(() => ({
    name,
    value,
    onChange
  }), [name, value, onChange]);
  return /*#__PURE__*/React.createElement("li", {
    role: "presentation"
  }, /*#__PURE__*/React.createElement("ul", Object.assign({
    role: "group",
    "aria-label": ariaLabel || name || 'Radio group'
  }, restProps, {
    className: bem(menuClass, radioGroupClass)(className),
    style: flatStyles(styles)
  }), /*#__PURE__*/React.createElement(RadioGroupContext.Provider, {
    value: contextValue
  }, children)));
}), 'MenuRadioGroup');
MenuRadioGroup.propTypes = { ...stylePropTypes(),
  'aria-label': PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.any,
  children: PropTypes.node.isRequired,
  onChange: PropTypes.func
};

export { ControlledMenu, FocusableItem, Menu, MenuButton, MenuDivider, MenuGroup, MenuHeader, MenuItem, MenuRadioGroup, SubMenu, applyHOC, applyStatics, useMenuState };
